pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,234 @@
+
+type workflowExecutor struct {
+	db                   repositories.RepositoryInterface
+	config               runtimeInterfaces.Configuration
+	executionManager     mgInterfaces.ExecutionInterface
+	snapshot             flyteSchedulerInterfaces.Snapshot
+	snapShotReaderWriter flyteSchedulerInterfaces.SnapshotReaderWriter
+	goGfInterface        flyteSchedulerInterfaces.GoGFWrapper
+}
+
+func (w *workflowExecutor) CheckPointState(ctx context.Context) {
+	workflowExecConfig := w.config.ApplicationConfiguration().GetSchedulerConfig().WorkflowExecutorConfig.FlyteWorkflowExecutorConfig
+	fileName := workflowExecConfig.SnapshotFileName
+	for true {
+		f, err := os.OpenFile(fileName, os.O_WRONLY | os.O_CREATE, 0644)
+		defer func(f *os.File) {
+			err := f.Close()
+			if err != nil {
+				logger.Errorf(ctx, "unable to close the file %v after writing due to %v", fileName, err)
+			}
+		}(f)
+		// Just log the error.
+		if err != nil {
+			logger.Errorf(ctx, "unable to create a writable snapshot file %v due to %v", fileName, err)
+		}
+		err = w.snapShotReaderWriter.WriteSnapshot(f, w.snapshot)
+		// Just log the error
+		if err != nil {
+			logger.Errorf(ctx, "unable to write the snapshot to file %v due to %v", fileName, err)
+		}
+		time.Sleep(snapshotWriterSleepTime * time.Second)
+	}
+}
+
+func (w *workflowExecutor) CatchUpAllSchedules(ctx context.Context, schedules []models.SchedulableEntity, toTime time.Time) error {
+	for _, s := range schedules {
+		err := w.CatchUpSingleSchedule(ctx, s, toTime)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+
+func (w *workflowExecutor) CatchUpSingleSchedule(ctx context.Context, s models.SchedulableEntity, toTime time.Time) error {
+	nameOfSchedule := w.goGfInterface.GetScheduleName(s)
+	lastExecTime := w.snapshot.GetLastExecutionTime(nameOfSchedule)
+	// Get the jitter value
+	workflowExecConfig := w.config.ApplicationConfiguration().GetSchedulerConfig().WorkflowExecutorConfig.FlyteWorkflowExecutorConfig
+	jitter := time.Duration(workflowExecConfig.JitterValue) * time.Second
+	schedulerEpochTime := workflowExecConfig.SchedulerEpochTime
+
+	var fromTime time.Time
+	// If the scheduler epoch time ise set then use that to catch up till the current time
+	// else check the last exec time for the schedule and if that also is not set then use current time with
+	// configured jitter
+	if !schedulerEpochTime.IsZero() {
+		fromTime = schedulerEpochTime
+	} else {
+		if !lastExecTime.IsZero() {
+			fromTime = lastExecTime
+		} else {
+			fromTime = time.Now().Add(-jitter)
+		}
+	}
+
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s, catchupTime, err)
```

Comment:
This needs to be abort instead of continuing further

Reply:
also would be good to add a sleep between calls to fire to avoid overwhelming admin/k8s

Reply:
I have added a rate limiter to admin which defines how many requests per sec should be sent to the admin

Reply:
https://github.com/uber-go/ratelimit

Reply:
Done